pkg/controllers/dnspolicy/dnspolicy_dnsrecords.go

```diff
@@ -137,3 +135,61 @@ func (r *DNSPolicyReconciler) deleteGatewayDNSRecords(ctx context.Context, gatew
 	}
 	return nil
 }
+
+func (r *DNSPolicyReconciler) buildClusterGateway(ctx context.Context, upstreamGateway *gatewayv1beta1.Gateway, downstreamCluster string) (dns.ClusterGateway, error) {
+	var target dns.ClusterGateway
+
+	mc := &clusterv1.ManagedCluster{}
+	if err := r.Client().Get(ctx, client.ObjectKey{Name: downstreamCluster}, mc, &client.GetOptions{}); err != nil {
+		return target, err
+	}
+
+	for _, address := range upstreamGateway.Status.Addresses {
+		if strings.Contains(address.Value, downstreamCluster) {
+			var gatewayAddresses []gatewayv1beta1.GatewayAddress
+
+			addressType := gatewayv1beta1.IPAddressType
+			if strings.Contains(string(*address.Type), string(gateway.MultiClusterHostnameAddressType)) {
+				addressType = gatewayv1beta1.HostnameAddressType
+			}
+
+			tmp := strings.Split(address.Value, "/")
+			addressValue := tmp[0]
+
+			if len(tmp) > 0 {
+				addressValue = tmp[len(tmp)-1]
+			}
+			gatewayAddresses = append(gatewayAddresses, gatewayv1beta1.GatewayAddress{
+				Type:  &addressType,
+				Value: addressValue,
+			})
+			target = *dns.NewClusterGateway(mc, gatewayAddresses)
+		}
+	}
+	return target, nil
+}
+
+func getGatewayAddresses(upstreamGateway *gatewayv1beta1.Gateway) []string {
+	var clusters []string
+
+	for _, address := range upstreamGateway.Status.Addresses {
+		value := strings.Split(address.Value, "/")
+		if strings.Contains(string(*address.Type), string(gateway.MultiClusterHostnameAddressType)) || strings.Contains(string(*address.Type), string(gateway.MultiClusterIPAddressType)) {
+			clusters = append(clusters, value[0])
+		}
+	}
+
+	return clusters
+}
```

Comment:
I think this would probably be better changed to return a map of clusters and addresses by moving your address processing logic into here:

```
func getClusterGatewayAddresses(gw *gatewayv1beta1.Gateway) map[string][]gatewayv1beta1.GatewayAddress {
	var clusterAddrs map[string][]gatewayv1beta1.GatewayAddress

	for _, address := range gw.Status.Addresses {
		var gatewayAddresses []gatewayv1beta1.GatewayAddress

		addressType := gatewayv1beta1.IPAddressType
		if *address.Type == gateway.MultiClusterHostnameAddressType {
			addressType = gatewayv1beta1.HostnameAddressType
		}

		//Default to Single Cluster (Normal Gateway Status)
		cluster := "none"
		addressValue := address.Value

		//Check for Multi Cluster (MGC Gateway Status)
		if *address.Type == gateway.MultiClusterIPAddressType || *address.Type == gateway.MultiClusterHostnameAddressType {
			tmp := strings.Split(address.Value, "/")
			//If this fails something is wrong and the value hasn't been set correctly
			if len(tmp) == 2 {
				cluster = tmp[0]
				addressValue = tmp[1]
			}
		}

		gatewayAddresses = append(gatewayAddresses, gatewayv1beta1.GatewayAddress{
			Type:  &addressType,
			Value: addressValue,
		})
		clusterAddrs[cluster] = gatewayAddresses
	}

	return clusterAddrs
}
```

Reply:
refactored the function